FIAT/finite_element.py

```diff
@@ -27,6 +28,82 @@
 
 
 class FiniteElement(object):
+    """Class implementing a basic abstraction template for special
+    types of finite elements, like traces. Most special methods will
+    need to be implemented by the special element in its module."""
+
+    def __init__(self, ref_el, order, formdegree=None, mapping="affine"):
+
+        self.order = order
+        self.formdegree = formdegree
+        self.mapping = mapping
+
+        self.ref_el = ref_el
+
+    def degree(self):
+        "Return the degree of the (embedding) polynomial space."
+        raise NotImplementedError()
```

Comment:
Your comments have been immensely helpful. I will push a change in a few moments that cuts a majority of this out. From the discussions I've had with David, `FiniteElement` is not meant to be used on it's own - it must have a proper implemented class that inherits from `FiniteElement`.

What I felt was bare minimum for this element is simply a `ReferenceElement` of some kind, the `order`, and a `formdegree` if one wants to do some FEEC magic at some point. Other properties/methods that are included in `CiarletElement` rely on specifying a `dual`, where it's used to performs things like construct the Vandermonde system, get `entity_dofs`, etc.

If you have more ideas for `FiniteElement`, I would be more than happy to discuss them with you next week.